### PR TITLE
end package cycles

### LINKED ISSUE #4645: package dependency graph has cycles
## Describe the bug
Running a Lerna command warns,
```
lerna notice cli v3.20.2
lerna WARN ECYCLE Dependency cycles detected, you should fix these!
lerna WARN ECYCLE @agoric/swingset-vat -> @agoric/notifier -> @agoric/swingset-vat
lerna WARN ECYCLE @agoric/deploy-script-support -> @agoric/vats -> @agoric/pegasus -> @agoric/deploy-script-support
lerna WARN ECYCLE @agoric/governance -> @agoric/zoe -> @agoric/governance
lerna WARN ECYCLE @agoric/run-protocol -> (nested cycle: @agoric/deploy-script-support -> @agoric/vats -> @agoric/pegasus -> @agoric/deploy-script-support) -> @agoric/run-protocol
lerna WARN ECYCLE @agoric/store -> (nested cycle: @agoric/swingset-vat -> @agoric/notifier -> @agoric/swingset-vat) -> @agoric/store
lerna WARN ECYCLE @agoric/swing-store -> @agoric/swing-store
lerna WARN ECYCLE @agoric/wallet-backend -> (nested cycle: @agoric/run-protocol -> (nested cycle: @agoric/deploy-script-support -> @agoric/vats -> @agoric/pegasus -> @agoric/deploy-script-support) -> @agoric/run-protocol) -> @agoric/wallet-backend
lerna WARN ECYCLE (nested cycle: @agoric/store -> (nested cycle: @agoric/swingset-vat -> @agoric/notifier -> @agoric/swingset-vat) -> @agoric/store) -> (nested cycle: @agoric/store -> (nested cycle: @agoric/swingset-vat -> @agoric/notifier -> @agoric/swingset-vat) -> @agoric/store)
lerna WARN ECYCLE (nested cycle: @agoric/wallet-backend -> (nested cycle: @agoric/run-protocol -> (nested cycle: @agoric/deploy-script-support -> @agoric/vats -> @agoric/pegasus -> @agoric/deploy-script-support) -> @agoric/run-protocol) -> @agoric/wallet-backend) -> (nested cycle: @agoric/wallet-backend -> (nested cycle: @agoric/run-protocol -> (nested cycle: @agoric/deploy-script-support -> @agoric/vats -> @agoric/pegasus -> @agoric/deploy-script-support) -> @agoric/run-protocol) -> @agoric/wallet-backend)
```

## To Reproduce
1. `yarn lerna --reject-cycles run lint`

## Expected behavior
No cyclic dependencies among packages.

## Prioritization Analysis

This problem is not catastrophic nor urgent. Here are the reasons to do it anyway:

1. Friction in development

The cycles in the dependency graph muddy the code architecture. It's not always clear where to put something and we end up moving things around as the boundaries become apparent. This suggests a lack of overall architectural clarity. We also sometimes hack around the irrational boundaries, such as https://github.com/Agoric/agoric-sdk/pull/5232

2. Prevents CI optimization of skipping unchanged packages

We can speed up CI and pre-push testing by only examining what has changed. For example, `lerna run test --since=main` will only run tests on packages that have changed since master. With cycles in the dependency graph it infers everything changed.

This is less relevant in CI where we run everything in parallel. We waste work but we don't waste wall time except for jobs like `deployment-test` which is very long and could automatically be skipped when it's known not to be affected.

3. Complicates semver

When concerns are mixed in a package, a breaking change to one part implies a breaking change to the whole package. 

4. Larger bundles

Because the Endo bundler don't yet have tree shaking, the lax package factoring can cause unnecessarily large bundles.

5. Unnecessary version bumps

Lerna bumps a version when its dependency bumps. 

```[tasklist]
### Tasks
- [ ] https://github.com/Agoric/agoric-sdk/issues/8088
```

### PR DESCRIPTION
**Pull request by turadg**
closes: #4645 

## Description

Eliminates all cycles in package dependency graph. Also enforces it by updating the `lerna` command to `--reject-cycles`.

Note that some undeclared dependencies remain. Solving those is deferred to https://github.com/Agoric/agoric-sdk/issues/8178

Stops publishing the `@agoric/boot` package which nothing outside agoric-sdk should have a need to import.

This also cleans up some types and other things in the course of working on the primary goal.

### Security Considerations

No change.

### Scaling Considerations

No changes.

### Documentation Considerations

Improves rationality of package dependencies. I don't know any documentation that is affected. If any is, I want to let it be discovered and get a CI test.

### Testing Considerations

CI

### Upgrade Considerations

Does not affect the runtime dependencies of contracts. Just moves testing and build dependencies.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -68,8 +68,8 @@
     "lint:format": "yarn prettier --check .github golang packages",
     "lint-fix": "yarn lerna run --no-bail lint-fix",
     "lint": "run-s --continue-on-error lint:*",
-    "lint:packages": "yarn lerna run --no-bail lint",
-    "test": "yarn lerna run --no-bail test",
+    "lint:packages": "yarn lerna --reject-cycles run --no-bail lint",
+    "test": "yarn lerna --reject-cycles run --no-bail test",
```
Comment (michaelfig) [inline package.json]: If you want to enforce rejection of cycles for all `lerna run` commands (do you really?), you can add a section to `lerna.json`:
```diff
diff --git a/lerna.json b/lerna.json
index a6b094cfe..61e5770e3 100644
--- a/lerna.json
+++ b/lerna.json
@@ -1,5 +1,8 @@
 {
   "command": {
+    "run": {
+      "rejectCycles": true
+    },
     "version": {
       "message": "chore(release): publish",
       "conventionalCommits": true,
```

File: packages/governance/test/swingsetTests/committeeBinary/bootstrap.js [hunk 2]
```diff
@@ -1,15 +1,15 @@
-import { E } from '@endo/eventual-send';
-import { Far } from '@endo/marshal';
-import { makeFakeBoard } from '@agoric/vats/tools/board-utils.js';
 import { makeMockChainStorageRoot } from '@agoric/internal/src/storage-test-utils.js';
 import buildManualTimer from '@agoric/zoe/tools/manualTimer.js';
+import { E } from '@endo/eventual-send';
+import { Far } from '@endo/marshal';
```
Comment (michaelfig) [inline packages/governance/test/swingsetTests/committeeBinary/bootstrap.js:4]: While you're changing this, you can use the preferred form:
```suggestion
import { E, Far } from '@endo/far';
```

File: packages/inter-protocol/test/smartWallet/test-oracle-integration.js [hunk 3]
```diff
@@ -53,7 +53,7 @@ const makeTestSpace = async (log, bundleCache) => {
     },
     psmParams,
   );
-  void psmVatRoot.bootstrap(...mockPsmBootstrapArgs(log));
+  void psmVatRoot.bootstrap(...mockPsmBootstrapArgs());
```
Comment (michaelfig) [inline packages/inter-protocol/test/smartWallet/test-oracle-integration.js:56]: I think this deserves a comment to explain why `log` isn't being forwarded.  Or maybe I'm missing something obvious?

File: packages/vats/test/devices.js
Comment (michaelfig) [inline packages/vats/test/devices.js:1]: Wow.  Nice! 🥇

**Review by michaelfig — APPROVED** [review]
LGTM!  Just a few comments, but none block landing this.

File: package.json (hunk 1 above)
Comment (michaelfig) [reply package.json]: Oh, and you can also add:
```js
    "exec": {
      "rejectCycles": true
    },
```

Comment (michaelfig) [reply package.json]: I'm just silly.  Here's a blanket way:
```diff
diff --git a/lerna.json b/lerna.json
index a6b094cfe..128122483 100644
--- a/lerna.json
+++ b/lerna.json
@@ -13,6 +13,7 @@
     "packages/*",
     "packages/wallet/*"
   ],
+  "rejectCycles": true,
   "useWorkspaces": true,
   "version": "independent"
 }
```

File: packages/governance/test/swingsetTests/committeeBinary/bootstrap.js (hunk 2 above)
Comment (turadg) [reply packages/governance/test/swingsetTests/committeeBinary/bootstrap.js:4]: There are over a hundred files with the two lines here. If `far` is preferred, I'll do that in a follow-up PR.

File: packages/inter-protocol/test/smartWallet/test-oracle-integration.js (hunk 3 above)
Comment (turadg) [reply packages/inter-protocol/test/smartWallet/test-oracle-integration.js:56]: With the refactor the receiving function already has it in scope.https://github.com/Agoric/agoric-sdk/blob/d751e4fcd94a7e3d5b94c80abab81bf23bb8ff71/packages/inter-protocol/test/smartWallet/contexts.js#L35-L37

File: packages/governance/test/swingsetTests/committeeBinary/bootstrap.js (hunk 2 above)
Comment (turadg) [reply packages/governance/test/swingsetTests/committeeBinary/bootstrap.js:4]: https://github.com/Agoric/agoric-sdk/pull/8180

**Pull request merged**